### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by tiserett**
[DEMO LINK](https://tiserett.github.io/react_todo-app-add-and-delete/)

### CODE REVIEW

**Review by Githi54 — APPROVED** [review]
Good job!🦾

File: src/App.tsx [hunk 1]
```diff
@@ -1,223 +1,176 @@
 /* eslint-disable jsx-a11y/control-has-associated-label */
```
Comment (AlinaMatuschak) [inline src/App.tsx:1]: don't disable eslint

File: src/App.tsx [hunk 2]
```diff
@@ -1,223 +1,176 @@
...
+  const [title, setTitle] = useState('');
+  const [isAdding, setIsAdding] = useState(false);
+  const [selectedTodos, setSelectedTodos] = useState<number[]>([]);
+
+  if (error) {
+    setTimeout(() => {
+      setError(null);
+      setIsAdding(false);
+    }, 3000);
+  }
```
Comment (AlinaMatuschak) [inline src/App.tsx:34]: you need to create setTimeout on the first render ErrorNotification component and clear timeout on unmount

File: src/App.tsx [hunk 3]
```diff
@@ -1,223 +1,176 @@
...
+        return todo.completed;
+
+      default:
+        return null;
+    }
+  });
 
   useEffect(() => {
     // focus the element with `ref={newTodoField}`
+    getTodos(userId)
```
Comment (AlinaMatuschak) [inline src/App.tsx:60]:
```suggestion
  const filteredTodos = todos.filter(todo => {
    switch (filterType) {
      case FilterType.All:
        return todo;

      case FilterType.Active:
        return !todo.completed;

      case FilterType.Completed:
        return todo.completed;

      default:
        return null;
    }
  });

  useEffect(() => {
    getTodos(user?.id || 0)
```

File: src/App.tsx [hunk 4]
```diff
@@ -1,223 +1,176 @@
...
+    }, 3000);
+  }
+
+  let userId = 0;
+
+  if (user?.id) {
+    userId = user?.id;
+  }
+
+  const filteredTodos = todos.filter(todo => {
```
Comment (AlinaMatuschak) [inline src/App.tsx:42]: useMemo

File: src/App.tsx [hunk 5]
```diff
@@ -1,223 +1,176 @@
...
+import { FilterType } from './types/FilterStatus';
+import { Todo } from './types/Todo';
+import { Error } from './types/Error';
 
 export const App: React.FC = () => {
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const user = useContext(AuthContext);
   const newTodoField = useRef<HTMLInputElement>(null);
+  const [error, setError] = useState<Error | null>(null);
+  const [filterType, setFilterType] = useState('all');
```
Comment (AlinaMatuschak) [inline src/App.tsx:23]:
```suggestion
  const [filterType, setFilterType] = useState(FilterType.All);
```

File: src/App.tsx [hunk 6]
```diff
@@ -1,223 +1,176 @@
...
+        return todo;
+
+      case FilterType.Active:
+        return !todo.completed;
+
+      case FilterType.Completed:
+        return todo.completed;
+
+      default:
+        return null;
```
Comment (AlinaMatuschak) [inline src/App.tsx:54]:
```suggestion
      case FilterType.Active:
        return !todo.completed;

      case FilterType.Completed:
        return todo.completed;

      default:
        return true;
```

File: src/App.tsx [hunk 7]
```diff
@@ -1,223 +1,176 @@
...
+    }
+  });
 
   useEffect(() => {
     // focus the element with `ref={newTodoField}`
+    getTodos(userId)
+      .then(setTodos)
+      .catch(() => setError(Error.LOADING));
```
Comment (AlinaMatuschak) [inline src/App.tsx:62]: Let’s use async await. It gives you a simpler way to work with asynchronous promise-based code

File: src/App.tsx [hunk 8]
```diff
@@ -1,223 +1,176 @@
...
+
     if (newTodoField.current) {
       newTodoField.current.focus();
     }
   }, []);
 
+  const handleSubmit = async (event: FormEvent) => {
```
Comment (AlinaMatuschak) [inline src/App.tsx:69]: useCallback

File: src/App.tsx [hunk 9]
```diff
@@ -1,223 +1,176 @@
...
+    removeTodo(todoId)
+      .then(() => {
+        setTodos([...todos.filter(todo => todo.id !== todoId)]);
+      })
+      .catch(() => {
+        setError(Error.DELETING);
+      });
+  };
+
+  const completedTodos = todos.filter(todo => todo.completed);
```
Comment (AlinaMatuschak) [inline src/App.tsx:104]: useMemo

File: src/App.tsx [hunk 10]
```diff
@@ -1,223 +1,176 @@
...
+
+  const deleteCompletedTodos = () => {
```
Comment (AlinaMatuschak) [inline src/App.tsx:106]: useCallback

File: src/components/ErrorNotification/ErrorNotification.tsx [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+}) => {
+  return (
+    <div
+      data-cy="ErrorNotification"
+      className={classnames(
+        'notification is-danger is-light has-text-weight-normal',
+        { hidden: !errorText },
+      )}
+    >
+      <>
```
Comment (AlinaMatuschak) [inline src/components/ErrorNotification/ErrorNotification.tsx:21]: is fragment necessary here?

File: src/components/Footer/Footer.tsx [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+};
+
+export const Footer: React.FC<Props> = ({
+  filterType,
+  handleFilterType,
+  todos,
+  deleteCompleted,
+}) => {
+  const todosLeft = todos.filter(todo => !todo.completed).length;
+  const todosCompleted = todos.filter(todo => todo.completed).length;
```
Comment (AlinaMatuschak) [inline src/components/Footer/Footer.tsx:19]: useMemo

File: src/components/Footer/Footer.tsx [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+      </nav>
+
+      <button
+        data-cy="ClearCompletedButton"
+        type="button"
+        className="todoapp__clear-completed"
+        onClick={deleteCompleted}
+      >
+        {todosCompleted > 0 && 'Clear completed'}
+      </button>
```
Comment (AlinaMatuschak) [inline src/components/Footer/Footer.tsx:71]: the empty button on page if I don’t have completed todos 🥲
Screenshot 2022-10-03 at 23 26 46

File: src/types/Error.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
+export enum Error {
```
Comment (AlinaMatuschak) [inline src/types/Error.ts:1]: 🔥

**Review by AlinaMatuschak — CHANGES_REQUESTED** [review]
Good work 👌

File: src/components/Footer/Footer.tsx (hunk 13 above)
Comment (tiserett) [reply src/components/Footer/Footer.tsx:65]: if I don't add empty button my filters gonna jump from left to the right when I finish any of the todos, if I add this empty button it reduces the amount of pixels they move and it slightly fixes the problem

File: src/components/ErrorNotification/ErrorNotification.tsx (hunk 11 above)
Comment (tiserett) [reply src/components/ErrorNotification/ErrorNotification.tsx:21]: it wasn't working before, but now it's working for some reason

File: src/components/Footer/Footer.tsx (hunk 13 above)
Comment (AlinaMatuschak) [reply src/components/Footer/Footer.tsx:65]: can you check how it works on mate demo?

Comment (AlinaMatuschak) [reply src/components/Footer/Footer.tsx:68]: how it looks on [mate demo link](https://mate-academy.github.io/react_todo-app-with-api/):
Screenshot 2022-10-05 at 01 01 35

**Review by AlinaMatuschak — APPROVED** [review]
Nice 👍 . But I believe that you will fix my comments on the next work )

**Pull request closed without merging**